File: mobile/src/main/java/com/siliconlabs/bledemo/ble/BluetoothDeviceInfo.java
package com.siliconlabs.bledemo.ble;

import android.bluetooth.BluetoothDevice;
import android.util.Log;

import com.siliconlabs.bledemo.beaconutils.BleFormat;

import java.util.ArrayList;

public class BluetoothDeviceInfo implements Cloneable {
    public BluetoothDevice device;
    public boolean isOfInterest;
    public boolean isNotOfInterest;
    public boolean serviceDiscoveryFailed;
    private BleFormat bleFormat = null;
    protected boolean hasAdvertDetails;
    public static int MAX_EXTRA_DATA = 3;
    private boolean connected;

    Object gattHandle;
    public boolean areServicesBeingDiscovered;

    public ScanResultCompat scanInfo;

    public BluetoothDeviceInfo() {

    }
    public boolean hasUnknownStatus() {
        return (!serviceDiscoveryFailed && !isNotOfInterest && !isOfInterest);
    }

    boolean isUnDiscovered() {
        return (gattHandle == null) && hasUnknownStatus();
    }

    void discover(BluetoothLEGatt bluetoothLEGatt) {
        serviceDiscoveryFailed = isNotOfInterest = isOfInterest = false;
        gattHandle = bluetoothLEGatt;
    }

    @Override
    public BluetoothDeviceInfo clone()  {
        final BluetoothDeviceInfo retVal;
        try {
            retVal = (BluetoothDeviceInfo)super.clone();
            retVal.device = device;
            retVal.scanInfo = scanInfo;
            retVal.isOfInterest = isOfInterest;
            retVal.isNotOfInterest = isNotOfInterest;
            retVal.serviceDiscoveryFailed = serviceDiscoveryFailed;
            retVal.bleFormat = bleFormat;
            retVal.gattHandle = null;
            retVal.connected = connected;
            retVal.hasAdvertDetails = hasAdvertDetails;
            retVal.areServicesBeingDiscovered = areServicesBeingDiscovered;
            return retVal;
        } catch (CloneNotSupportedException e) {
            Log.e("clone","Could not clone" + e);
        }
        return null;
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof BluetoothDeviceInfo)) {
            return false;
        }

        final BluetoothDeviceInfo that = (BluetoothDeviceInfo) o;
        return device.equals(that.device) && (isOfInterest == that.isOfInterest) && (isNotOfInterest == that.isNotOfInterest) && (serviceDiscoveryFailed == that.serviceDiscoveryFailed);
    }

    @Override
    public int hashCode() {
        return device.hashCode();
    }

    @Override
    public String toString() {
        return scanInfo.toString();
    }

    public BleFormat getBleFormat() {
        if (bleFormat == null) {
            bleFormat = BleFormat.getFormat(this);
        }
        return bleFormat;
    }

    public void setBleFormat(BleFormat bleFormat) {
        this.bleFormat = bleFormat;
    }

    public void setAdvertData(ArrayList<String> advertisements) {
        if (scanInfo != null) {
            scanInfo.setAdvertData(advertisements);
        }
    }

    public ArrayList<String> getAdvertData() {
        if (scanInfo != null) {
            return scanInfo.getAdvertData();
        }
        return new ArrayList<>();
    }

    public boolean hasAdvertDetails() {
        return hasAdvertDetails || getAdvertData().size() > MAX_EXTRA_DATA;
    }
    public boolean isConnected() {
        return connected;
    }

    public void setConnected(boolean connected) {
        this.connected = connected;
    }

    public int getRssi() {
        return scanInfo.getRssi();
    }

    public void setRssi(int rssi) {
        this.scanInfo.setRssi(rssi);
    }

    public String getName() {
        return device.getName();
    }

    public String getAddress() {
        return device.getAddress();
    }
}

File: mobile/src/main/java/com/siliconlabs/bledemo/Bluetooth/BLE/BluetoothDeviceInfo.kt
package com.siliconlabs.bledemo.bluetooth.ble

import android.bluetooth.BluetoothDevice
import android.util.Log
import androidx.core.util.isEmpty
import com.siliconlabs.bledemo.bluetooth.beacon_utils.BleFormat
import com.siliconlabs.bledemo.bluetooth.beacon_utils.BleFormat.Companion.getFormat
import java.util.*
import kotlin.math.min

class BluetoothDeviceInfo(var device: BluetoothDevice, var isFavorite: Boolean = false) : Cloneable {

    var connectionState = ConnectionState.DISCONNECTED
    var isConnectable = false

    var bleFormat: BleFormat? = null
    var scanInfo: ScanResultCompat? = null
    var rawData: String? = null

    var intervalNanos = 0L
    var count = 0
    var timestampLast: Long = 0



    public override fun clone(): BluetoothDeviceInfo {
        val retVal: BluetoothDeviceInfo
        try {
            retVal = super.clone() as BluetoothDeviceInfo
            retVal.device = device
            retVal.scanInfo = scanInfo
            retVal.bleFormat = bleFormat
            retVal.connectionState = connectionState
            retVal.isConnectable = isConnectable
            retVal.isFavorite = isFavorite
            retVal.rawData = rawData
            return retVal
        } catch (e: CloneNotSupportedException) {
            Log.e("clone", "Could not clone$e")
        }
        return BluetoothDeviceInfo(device)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is BluetoothDeviceInfo) {
            return false
        }
        return device == other.device
    }

    override fun hashCode(): Int {
        return device.hashCode()
    }

    override fun toString(): String {
        return scanInfo.toString()
    }

    fun getBleFormat(shouldCheckAgain: Boolean): BleFormat {
        if (bleFormat == null || shouldCheckAgain) {
            // device can be programmed to advertise changing data, switching between BLE formats
            bleFormat = getFormat(this)
        }
        return bleFormat!!
    }

    fun setIntervalIfLower(intervalNanos: Long) {
        if (intervalNanos <= 0L) return
        if (this.intervalNanos == 0L) this.intervalNanos = intervalNanos
        else if (intervalNanos < this.intervalNanos * 0.7 && count < 10)
            this.intervalNanos = intervalNanos
        else if (intervalNanos < this.intervalNanos + 3000000) {
            val limitedCount = min(count, 10)
            this.intervalNanos = (this.intervalNanos * (limitedCount - 1) + intervalNanos) / limitedCount
        } else if (intervalNanos < this.intervalNanos * 1.4) {
            this.intervalNanos = (this.intervalNanos * 29 + intervalNanos) / 30
        }
    }


    val advertData: ArrayList<String?>
        get() = scanInfo?.advertData ?: arrayListOf()

    var rssi: Int
        get() = scanInfo?.rssi ?: 0
        set(rssi) {
            scanInfo?.rssi = rssi
        }

    val name: String
        get() = device.name ?: "N/A"

    val address: String
        get() = device.address

    val manufacturer: DeviceManufacturer
        get() = scanInfo?.scanRecord?.manufacturerSpecificData?.let {
            if (it.isEmpty()) DeviceManufacturer.UNKNOWN
            else when (it.keyAt(0)) {
                MANUFACTURER_VALUE_WINDOWS -> DeviceManufacturer.WINDOWS
                else -> DeviceManufacturer.UNKNOWN
            }
        } ?: DeviceManufacturer.UNKNOWN

    enum class DeviceManufacturer {
        WINDOWS,
        UNKNOWN
    }

    enum class ConnectionState {
        CONNECTED,
        CONNECTING,
        DISCONNECTED
    }

    companion object {
        private const val MANUFACTURER_VALUE_WINDOWS = 6
    }

}